runtime/parachains/src/scheduler.rs

```diff
@@ -13,3 +13,1403 @@
+		/// parathread-multiplexers.
+		///
+		/// Bounded by the number of cores: one for each parachain and parathread multiplexer.
+		AvailabilityCores: Vec<Option<CoreOccupied>>;
+		/// An index used to ensure that only one claim on a parathread exists in the queue or is
+		/// currently being handled by an occupied core.
+		///
+		/// Bounded by the number of parathread cores and scheduling lookahead. Reasonably, 10 * 50 = 500.
+		ParathreadClaimIndex: Vec<ParaId>;
+		/// The block number where the session start occurred. Used to track how many group rotations have occurred.
+		SessionStartBlock: T::BlockNumber;
+		/// Currently scheduled cores - free but up to be occupied. Ephemeral storage item that's wiped on finalization.
+		///
+		/// Bounded by the number of cores: one for each parachain and parathread multiplexer.
+		Scheduled get(fn scheduled): Vec<CoreAssignment>; // sorted ascending by CoreIndex.
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> { }
+}
+
+decl_module! {
+	/// The scheduler module.
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin {
+		type Error = Error<T>;
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Called by the initializer to initialize the scheduler module.
+	pub(crate) fn initializer_initialize(_now: T::BlockNumber) -> Weight {
+		Self::schedule(Vec::new());
+
+		0
+	}
+
+	/// Called by the initializer to finalize the scheduler module.
+	pub(crate) fn initializer_finalize() {
+		// Free all scheduled cores and return parathread claims to queue, with retries incremented.
+		let config = <configuration::Module<T>>::config();
+		ParathreadQueue::mutate(|queue| {
+			for core_assignment in Scheduled::take() {
+				if let AssignmentKind::Parathread(collator, retries) = core_assignment.kind {
+					let entry = ParathreadEntry {
+						claim: ParathreadClaim(core_assignment.para_id, collator),
+						retries: retries + 1,
+					};
+
+					if entry.retries > config.parathread_retries { continue }
+					queue.queue_entry(entry, config.parathread_cores);
+				}
+			}
+		})
+
+	}
+
+	/// Called by the initializer to note that a new session has started.
+	pub(crate) fn initializer_on_new_session(notification: &SessionChangeNotification<T::BlockNumber>) {
+		let &SessionChangeNotification {
+			ref validators,
+			ref random_seed,
+			ref new_config,
+			..
+		} = notification;
+		let config = new_config;
+
+		let mut thread_queue = ParathreadQueue::get();
+		let n_parachains = <paras::Module<T>>::parachains().len() as u32;
+		let n_cores = n_parachains + config.parathread_cores;
+
+		<SessionStartBlock<T>>::set(<system::Module<T>>::block_number());
+		AvailabilityCores::mutate(|cores| {
+			// clear all occupied cores.
+			for maybe_occupied in cores.iter_mut() {
+				if let Some(CoreOccupied::Parathread(claim)) = maybe_occupied.take() {
+					let queued = QueuedParathread {
+						claim,
+						core_offset: 0, // this gets set later in the re-balancing.
```

Comment:
if there are times where this can be invalid, consider an `Option<u32>` instead, or split the type into two `*Floating` + `*Scheduled` with `From<*Floating> for *Scheduled` impl where `*` denotes `QueuedParathread`

Reply:
That sounds like it would get messy. If I made it into an `Option`, we'd have to `expect` the option in quite a few places.

Likewise if I made it an `enum` with `Floating` and `Scheduled`, which I think is what you're suggesting.

It's not that this is invalid, it's just unbalanced, and only in this specific section of code. And then it gets re-balanced right after, so I don't really see the issue.